File: modules/common/src/main/java/com/falsepattern/zigbrains/common/MultiConfigurable.java
/*
 * Copyright 2023-2024 FalsePattern
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.falsepattern.zigbrains.common;

import com.intellij.openapi.options.Configurable;
import com.intellij.openapi.options.ConfigurationException;
import com.intellij.ui.JBColor;
import lombok.val;
import org.jetbrains.annotations.Nullable;

import javax.swing.JComponent;
import java.util.Arrays;

import static com.falsepattern.zigbrains.common.util.dsl.JavaPanel.newPanel;

public abstract class MultiConfigurable implements Configurable {
    private final SubConfigurable[] configurables;
    protected MultiConfigurable(SubConfigurable... configurables) {
        this.configurables = Arrays.copyOf(configurables, configurables.length);
    }

    @Override
    public @Nullable JComponent createComponent() {
        return newPanel(p -> {
            for (val configurable: configurables) {
                configurable.createComponent(p);
            }
        });
    }

    @Override
    public boolean isModified() {
        for (val configurable: configurables) {
            if (configurable.isModified())
                return true;
        }
        return false;
    }

    @Override
    public void apply() throws ConfigurationException {
        for (val config: configurables) {
            config.apply();
        }
    }

    @Override
    public void reset() {
        for (val config: configurables) {
            config.reset();
        }
    }

    @Override
    public void disposeUIResources() {
        for (val config: configurables) {
            config.disposeUIResources();
        }
    }
}

File: core/src/main/kotlin/com/falsepattern/zigbrains/shared/MultiConfigurable.kt
/*
 * This file is part of ZigBrains.
 *
 * Copyright (C) 2023-2025 FalsePattern
 * All Rights Reserved
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *
 * ZigBrains is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, only version 3 of the License.
 *
 * ZigBrains is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with ZigBrains. If not, see <https://www.gnu.org/licenses/>.
 */

package com.falsepattern.zigbrains.shared

import com.intellij.openapi.options.Configurable
import com.intellij.openapi.util.Disposer
import com.intellij.ui.dsl.builder.panel
import javax.swing.JComponent

abstract class MultiConfigurable(private val configurables: List<SubConfigurable>): Configurable {
    override fun createComponent(): JComponent? {
        return panel {
            for (configurable in configurables) {
                configurable.createComponent(this)
            }
        }
    }

    override fun isModified(): Boolean {
        return configurables.any { it.isModified() }
    }

    override fun apply() {
        configurables.forEach { it.apply() }
    }

    override fun reset() {
        configurables.forEach { it.reset() }
    }

    override fun disposeUIResources() {
        configurables.forEach { Disposer.dispose(it) }
    }
}
